### DISCUSSION TITLE
IDEAS for Packages - Where's the best place to share some of my ideas?

### DISCUSSION THREAD

**Discussion by marvelloz** [PresearchOfficial/presearch-packages#173] [Ideas]
I'm not a programmer, however I do have deep experience working with Search Engines. I have been working in the Search Engine industry for over 2 decades. Just 2 of my many career highlights:
1) I work at Microsoft at the time they launched Bing. I was involved in many internal think tanks on how to beat Google.
2) Was later hired by Google Cloud as there inhouse Search & Digital Marketer. This gave me access to Matt Cutts (at the time) and network with the Google engineers on the search team.  

I have a passion and interest in making search engine interesting, fun, useful and profitable. 
I'm a HUGE fan of Presearch and I'm running a node to support this decentralised privacy-first project.  

Please let me know if this is where I should be sharing my ideas for packages or if there's a better forum?  

Thanks.

**Comment by jejopl** [comment]
Hey @mvozzo - yes, this is the best place for the ideas. We will highly appreciate it 😀 👍

**Comment by ghost** [comment]
Hi everyone and hi mvozzo.

I have 4 ideas that are complementary to beat google, being open source.

**List of ideas**
1. create research submodules: you can check this idea here [Have you all considered developing separate presearch packages in the repository?](https://github.com/PresearchOfficial/presearch-packages/discussions/174) example: **"PresearchOfficial\presearch-youtube"** etc - each repository within PresearchOfficial, is the site or url of the site to be accessed.
2. smart contract: Smart contracts allow for some automation. We can have, for example, smart contracts to automate which urls should be searchable, for example, we can establish that searchable urls must be public or private, or that they are user-verifiable or that they are in accordance with tos (terms of service), for example , avoid urls that contain material that is wrong in some country or law. Currently, I haven't seen any academic research being done on search engines with smart contracts. But I think it's a good idea initially, but it would require more academic research and user experiments to see if it's a really good idea for most people or all users.
3. Another way to make presearch more known and commercially used would be if presearch could store, as if it were a cdn, sitemaps. Every website or online content management system has mechanisms to generate a sitemap, a map that contains a list of urls. If the presearch is a cdn (content delivery network) this allows you to share all urls on the site in a distributed way. Which increases the chance of research.
4. WebQuery: You can read this proposal here: [rikmeijer/WebQuery](https://github.com/rikmeijer/WebQuery). For example, if you want your site to be presearch indexable, you would need to add this line of code to your site. That way, every search that the user does will be indexed in presearch. Example:

#### index-presearch-remote.html
```html
<div hx-target="this" hx-swap="outerHTML">
    <div><label for="site-search">Search the site or local... url-file:</label>:</div>
    <button hx-get="https://www.sample-presearch.com/api/presearch-search/url/add" class="btn btn-primary">
    Search <!--- Click to add site-search with api presearch-api-search --->
    </button>
</div>
```

**code-description:** In this code, all the search performed is sent by get to the presearch server. This allows you to have an international search API, so that everything is indexable automatically and in a decentralized way and distributed by presearch. 

*Just a thought.*

**Reply by marvelloz** [reply]
Hey @codehangen , great idea around Presearch branching out into the sphere of being a CDN and storage of sitemaps.  
Every site I worked on had a sitemap. Some small, some HUGE.  I have been involved in creating countless number of sitemap.xml files and these assisted Google/Bing in the indexing of content. 

Just to follow-on from Codehangen's idea...
If a service existed were website owners could pay a small amount in PRE tokens to Presearch to have them daily scrape and store a copy of their sitemap.xml to ensure their site remains fully indexed in Presearch's distributed database, that would be great. 
These PRE tokens could then be divided up and passed directly to the Presearch Nodes that are storing these sitemaps.

**Reply by ghost** [reply]
Hi mvozzo, thank you for feedback.

> Hey @codehangen , great idea around Presearch branching out into the sphere of being a CDN and storage of sitemaps.
Every site I worked on had a sitemap. Some small, some HUGE. I have been involved in creating countless number of sitemap.xml files and these assisted Google/Bing in the indexing of content.

Please, see this: [ekalinin/sitemap.js/issues/400](https://github.com/ekalinin/sitemap.js/issues/400), [bigskysoftware/htmx/discussions/1058](https://github.com/bigskysoftware/htmx/discussions/1058) ... (an note... I was talking about sitemaps and cdn in these links I provided here and how that might be interesting.)

*Just a thought.*

> If a service existed were website owners could pay a small amount in PRE tokens to Presearch to have them daily scrape and store a copy of their sitemap.xml to ensure their site remains fully indexed in Presearch's distributed database, that would be great.
These PRE tokens could then be divided up and passed directly to the Presearch Nodes that are storing these sitemaps.

Great idea.

**Comment by marvelloz** [comment]
Regarding @codehangen's idea on WebQuery. 

When I was working at Google Cloud (formerly Google Enterprise), one of the BIGGEST revenue generators for them at the time was a product called 'Enterprise Search' it has had a lot of different product name changes since then, but here's what it did. It was for use by businesses not wanting to host and maintain there own search infrastructure and instead relied on Google's infrastructure to power search on there websites. Google dedicated servers to the business and indexed there web content and powered the search experience.   

Now, with the growing distrust of BIG TECH, I think there is a window of opportunity where Presearch could market to these Enterprise users to shift across to an decentralized open source solution and not rely on Big Tech's spyware.

I like CodeHangen's idea very much. Even if the solution was just targetted to small businesses and offering them some way to easily implement the Presearch Search Engine on their own website for free.  This would be great marketing for Presearch as you could add in some tagline on the Search Results page  "Powered by a Free, Open Source, Decentralised Search Engine - Presearch.com"

**Reply by ghost** [reply]
Hi mvozzo, thank you for feedback.

> Now, with the growing distrust of BIG TECH, I think there is a window of opportunity where Presearch could market to these Enterprise users to shift across to an decentralized open source solution and not rely on Big Tech's spyware.

Great idea, feedback. So... therefore, I think whether presearch or [findto](https://github.com/lucasm/findto) can be interesting options for this.  (An note... so... I was thinking about working with the sitemap.js framework with smart contracts and ipfs to this... "decentralized open source" an... another note... there is an interesting project called [orbit-db](https://github.com/orbitdb/orbit-db) - which is a distributed p2p database. I wonder if there would be no need to use a distributed, decentralized database like orbit-db for presearch).

*Just a thought.*

> I like CodeHangen's idea very much. Even if the solution was just targetted to small businesses and offering them some way to easily implement the Presearch Search Engine on their own website for free. This would be great marketing for Presearch as you could add in some tagline on the Search Results page "Powered by a Free, Open Source, Decentralised Search Engine - Presearch.com"

thank you for feedback again.

**Comment by marvelloz** [comment]
### PACKAGE IDEA: Foreign Exchange Currency Converter

Google has a good FX exchange rate converter and charts module.  

I feel Presearch could create an even better one.  
There would be a HUGE opportunity for Presearch to create a similar package so that when someone searches for example:  EUR to USD
The package could contain:

- exchange rate 
- 7 day chart (can be adjusted to monthly ore yearly trend chart)
- Provide a calculator that has a default  [  100] EUR   =  [ XXXX] USD  , however, allow the user to modify the input fields whilst prominently display the latest exchange rate. 

**How you would monetise this package?**
This would be done through advertising the brand/logo to the highest PRE token staked for each given FX pair search phrase.

**Reply by ghost** [reply]
Hi mvozzo, great idea.

**Comment by ghost** [comment]
Hi everyone. 

#### PACKAGE IDEA: Peertube in Presearch (concept)
I have an interesting idea and want to hear feedback here. I wanted to have Peertube and any decentralized video application integrated into Presearch.

#### why Peertube in Presearch? 
1. If you think that most peertube users want a decentralized, distributed search engine. It would make sense to have Peertube support in Presearch... or even... thinking of another scenario... imagine that most Linux users want an alternative to Google search and Youtube, some of these users can use solutions like Libre Wolf, Mozilla FireFox as an alternative to Google Chrome or even Peertube as alternative to Youtube, or even findTo or presearch or duck-duck-go also as an alternative to Google Search... so... think that users would have no difficulty using an alternative solution to Google Search and Youtube - if only there was an integration of Peertube and Presearch. *My main argument that I want to convince is that there can be a familiarity of Peertube and Presearch for the largest number of people - if both projects are integrated or share the same interests.*
3. Both Peertube and Presearch are open source projects. That's a good reason, both projects have the same "commercial value or common goal". In this sense, imagining a partnership between two projects is not bad or strange. "Both want to be decentralized, distributed open solutions." *So, if there's a partnership between both projects, that's great for both projects. It's a win-win relationship - so... this can contribute to most people using both projects at the same time for the same purpose... instead of... using the two projects separately.*
4. Another reason if you think on a commercial level or common goal, the more users use Presearch and Peertube the more chance these solutions are popular, known or famous. And greater is the chance of number of developers, volunteers and partners. *It may not reduce most Google and YouTube users, but it's a start.*
5. And if you think that having integration is a good goal, integrating Peertube with Presearch would be great.
6. It's hard for an average user to use things that aren't popular. Home users use familiar things, when something integrated is provided, a set of applications becomes common to the home user and he starts to use it. So... for example... every "Linux distro" or "operating system" has a set of applications that the user will use. What I said can be applied to a "user interface" or "desktop environment" that has a set of rules, patterns and conventions for greater use. This is done for greater familiarity and also better ui/ux for home users. For example, are you going to research something? "google chrome and google search" or will you watch videos? "youtube"... etc... etc. *So... having Peertube and Presearch integration can be something interesting for most people who want an alternative to Google Search and at the same time to Youtube. So... this can create a certain familiarity that I mentioned or a good user experience.*
7. Also, both open source projects have their recognition, I don't see why there wouldn't be a need to not integrate the two with the same goal. For the end user, this would be a great starting point, having two projects rolled into one.


#### Why is this feature interesting?
- "open source apps on presearch": I think it's important to partner with several open companies or open and decentralized solutions. This allows for greater integration and knowledge of various open source applications. And it allows presearch to stand out against Google Search. It's an interesting thing to think about, as I haven't seen this feature in duck-duck-go or findTo. *All the searches I carry out are sent to presearch, so I have a local, remote, distributed and decentralized database.* 
- In the previous comment I talked about the importance of [Peertube](https://joinpeertube.org/) in Presearch, this same example could be applied to [mastodon](https://mastodon.social/explore) which is an alternative to Twitter or something [Discourse](https://www.discourse.org/) which is an alternative to GitHub Discussions for Presearch... 
- *So... "everything that is searched in Mastedon or Discourse is saved in the Presearch, and I can choose whether this information waits saved publicly or not."* Peertube, Mastedon or Discourse instances can tell me data constraints e.g. storage time, content type etc. 

#### How you would monetise this package?
The money comes from maintaining, improving and enhancing the code collaboratively with financial donation by developers, users or through a commercial agreement between different open source applications as discounts on their cloud plans.

*Just a thought.*

**Reply by marvelloz** [reply]
This is yet another great idea from @codehangen !

I'd like to add to this that Presearch should form alliances and partnerships with other major decentralised platforms to become the "Search Partner" for these projects.  

For example, please build out "site" search partnerships with the likes of (not a full comprehensive list): 
- LBRY ( https://odysee.com/ ) - Decentralised Video Platform
- Bastyon.com  ( https://bastyon.com ) - Decentralised Social Media Platform
- Theta TV ( https://thetatv.com ) - Decentralised Video Streaming Platform

I'm sure there are many other decentralised platforms that need Search functionality for users to discover content. 

**How to monetise?**
By establishing these partnerships, part of the arrangement would be for PreSearch to gain access to API's/XML feeds to assist Presearch with indexing content. The Presearch SERPs could then be customised to display Keyword stake ads in the results for related Keyword search terms. Once Presearch offers a CPC revenue model, they could offer rev share opportunities with the platform owner.

**Reply by ghost** [reply]
Hi mvozzo.

> This is yet another great idea from @codehangen !

thank you for feedback.

> By establishing these partnerships, part of the arrangement would be for PreSearch to gain access to API's/XML feeds to assist Presearch with indexing content. The Presearch SERPs could then be customised to display Keyword stake ads in the results for related Keyword search terms. Once Presearch offers a CPC revenue model, they could offer rev share opportunities with the platform owner.

Users could vote on the search pages if they wanted to... it's common for users to create curated lists of links, services, products. This would be a way to customize the user experience with something more organic, human. *(Money can be received or sent through curated links between one or more users in a decentralized, distributed way - What I'm talking about here can be complementary to your idea)*


*Just a thought.*

**Comment by ghost** [comment]
Hi everyone.

#### PACKAGE IDEA: decentralized presearch data (concept)
I would like to choose the storage provider for my presearch data. For example, I want to choose where I sync my registration data or any proprietary data I have to: *remote-storage(protocol), solid-project(protocol), rss3(protocol), webdav(protocol), ftp(protocol), [EteSync](https://www.etesync.com/)(protocol), [DAVx⁵](https://www.davx5.com/)(protocol), DecSync CC(protocol), CalDAV(protocol), hypercore(protocol), ipfs(protocol) etc... or mongodb(database), postgres(database) etc...*

#### image
![image](https://user-images.githubusercontent.com/111701513/206940651-53af3196-b6ac-4dde-875f-b790a50a859f.png)

**image-description:** As we can see in this image, we can choose where we are going to store our data, for this to be possible it is necessary that there is a greater integration of the data with network protocols such as: *remote-storage, solid-project, smdb, rss3, webdav, ftp, EteSync, DAVx⁵, DecSync CC, CalDAV, hypercore, ipfs etc.* or services such as: google drive, dropbox, trezorit etc... or... database such as: mongodb, postgres, orbitdb etc... - this way we have decentralized user data, data outside the server or the presearch application.

#### How you would monetise this package?
1. we may charge for maintenance of network protocols in presearch. That is, charge maintenance for data synchronization in the network protocol that the user, company or community wants.
2. we can have paid and free synchronized network protocols as plugins. like this... users can pay or provide network protocols... or they can use paid and free plugins... paid plugins... can be plugins that need donation... free plugins are provided and maintained in a way opened by the community.

#### Why is this feature interesting?
- App search data is maintained and provided by presearch. *"this can increase the commercial value of presearch for users, developers, companies and communities."*
- the user decides where to store, share his data on different servers, etc. Storage can be local via session, cookies, localstorage. Or it can be remote/local with remote-storage, ipfs etc.

#### How could this be done?
There are projects like [typeorm/typeorm](https://github.com/typeorm/typeorm) or things like [xo/usql](https://github.com/xo/usql) - which allow you to store, share various types of data in different databases. These projects can be used in the backend to serve as a bridge for synchronizing, transferring data to different network protocols, third-party services or databases.


*Just a thought.*

**Comment by ghost** [comment]
Hi everyone.

I would like to develop or contribute to a list of ideas or features that I would like the pre-search to have. I have an interesting idea and want to hear feedback here... I have 4 new ideas that are complementary to beat google, being open source.

*List of ideas or features I would like the pre-search to have*
- "import/export bookmarks in presearch": It's very common to save things like bookmarks in the browser or save bookmarks to the cloud. I would like to import my bookmarks into presearch to help with content indexing. I could get paid to categorize or categorize presearch links too.
- "rss(feed) in presearch": I would like to distribute a presearch search rss for others to follow what I search. This is interesting for academic researchers. *You could receive notifications of what people search for in a distributed, decentralized, secure and private way.*
- "get paid to curate links or pay to curate links": there are projects like [awesome-ipfs](https://awesome.ipfs.tech/) that list useful links for designers, ui-designers, ux-designers, programmers etc with ipfs. *This is a way for you to index relevant content without artificial intelligence or ads. I can pay some money or provide interesting links to people, and that way I also earn money. Or it could have benefits like open badges and discounts on plans in open source apps like notesnook, joplin among others.*
- Increase distribution channels: With this idea PresearchOfficial is accessed, delivered to all types of end users, developers. It is known that there are several package managers, my general idea is to provide a PresearchOfficial api for the channels of package managers in programming languages, they usually work independently of the operating system: windows, linux, mac etc.

#### concept: "Increase distribution channels"
![image](https://user-images.githubusercontent.com/111701513/207221101-6eb6bb0c-1745-46fe-b094-a15b01e49a77.png)

**image-description:**  For example...  [Ninite/PresearchOfficial](https://ninite.com/), [Chocolatey/PresearchOfficial](https://chocolatey.org/), [unpkg/PresearchOfficial](https://unpkg.com/) *this allows PresearchOfficial to be used for web app like spa, mpa, pwa, website etc.* Or [NuGet/PresearchOfficial](https://www.nuget.org/) *this allows PresearchOfficial to be used for web app like spa, mpa, pwa, website etc or desktop for C# applications.*

*Just a thought.*

**Comment by marvelloz** [comment]
**PACKAGE IDEA:  Presearch Polls / Surveys**

Data is the new "gold" in the digital age. This is why Google, Microsoft, Amazon are leaders in Big Tech.
 
I worked at a company called Qualtrics and they specialised in Experience Management. They had cloud based software for doing detailed market research and survey data analysis). There is BIG $money in this area and I feel Presearch could develop a Package to assist organisations conduct surveys. 

The first version of this Package should focus on improving Presearch's own search results. 

Here are 2 examples:  

**1) EXIT POLL**
Run an Exit Poll to rate the quality of the search experience.  

The polls/surveys would be optional (meaning the user would opt-in)  they would be 100% anonymous.  The user would be rewarded with some PRE tokens. 

Have some scales 1 - 10 to measure things like:
a) Rate the quality to the search results (1 = poor,  10 = spot on) 
b) How likely are you to use Presearch again  ( 1 = never again,   10 = definitely use again)  
c) perhaps include an open ended feedback textbox.
This data could be married up with the search keyword and sent anonymously to Presearch to improve the search results page. 
  
 
**2) PRESEARCH vs GOOGLE - FIGHT!**

This would be a Fun Game-like Survey that Presearch could allow users to Opt-in and earn some PRE.
The Game would compose of 10 Survey questions which would be based on 10 random search queries that that Presearch wanted to analyse and compare themselves to the Google. 

About the User Interface
Once the user Opts-in to participates in this Survey, a new tab opens that has fixed Top Nav which has the survey question and is where the user will submit their answers.  underneath this Top nav section the page is divided into 2 columns and would be like 2 iframes side-by-side.  

Once the used hits the "Start Fight" button, the first of 10 round of search queries are presented to the survey participant. 
In the left iframe would be the search results from Google,  on the right would be the search results from Presearch and the user would have to choose which SERP (Search Engine Results Page) provided the best results. 

In the top nav there would be the questions:
"Choose your winner..."  and above the first column is  a button "Google"   and above the 2nd column is "Presearch".
After the user selects a winner, the next fight round starts. 
But this time the columns are switched so that Google is not always presented in the left column.  This heaps to change things ups.  Perhaps adding some randomness to the presentation. 

Presearch can then use this data to test some of it's new "packages" and also test certain biased results according to Google and unbiased (shall we say uncensorsored) results under Presearch. So this tool can also educate users on the better quality of results from Presearch.  

The above would be just 2 idea for Presearch to consider, but I hope you can see that this package could be expanded to advertisers that want to carry our research on users as they are doing search. 

Just a thought.  

Hey @jejopl , please let me know your thoughts on this ideas from @codehangen and myself.

**Reply by ghost** [reply]
Hi mvozzo.

> PACKAGE IDEA: Presearch Polls / Surveys
> Data is the new "gold" in the digital age. This is why Google, Microsoft, Amazon are leaders in Big Tech.
> I worked at a company called Qualtrics and they specialised in Experience Management. They had cloud based software for doing detailed market research and survey data analysis). There is BIG $money in this area and I feel Presearch could develop a Package to assist organisations conduct surveys.
> The first version of this Package should focus on improving Presearch's own search results.

With orbit-db it is possible to receive and send certain p2p data, I think in this use case... if the presearch wants... to evaluate the searches or user experience in a p2p network this would be possible with a p2p database like orbit -db. Each network node can create new questions and answers... this would make the pre-survey more known, popular and usable - in a decentralized way, distributed with the mutual support of different people around the world.

You would need a presearch package that sends and receives questions and answers from different users. As well as, register and update new questions and answers. This could be done with a bounty program like bug-bounty where you receive money for features you develop. 

**Update:** An interesting idea would be to provide a search notification for presearch partner pages through the w3c [webmention](https://www.w3.org/TR/webmention/) protocol, which notifies people according to the shared page accessed. Note: *today there are things like radicle, git-bug, ipfs, hypercore, orbit-db etc that have these features that I mentioned here.*

> Hey @jejopl , please let me know your thoughts on this ideas from @codehangen and myself.

thank you for feedback.

**Comment by ghost** [comment]
Hi everyone.

#### concept
I was thinking about creating custom themes for the pre-search.

#### Presearch-theme-default *before*

![image](https://user-images.githubusercontent.com/111701513/207102460-50edd3fa-94aa-403f-b2a0-87248463a130.png)

#### Presearch-Mastodon-theme *after*
![image](https://user-images.githubusercontent.com/111701513/207077981-00ecf217-f81e-4c02-85e6-f9fb81491d25.png)

**image-description:** As we can see in the image, "I thought about improving the presearch search interface". This is a Mastodon custom theme demo for Presearch.

#### Presearch-GoogleSearch-theme *after*
![image](https://user-images.githubusercontent.com/111701513/207135830-45d67524-97aa-4bb6-9d2f-e1afa1876064.png)

**image-description:** As we can see in the image, "I thought about improving the presearch search interface". This is a GoogleSearch custom theme demo for Presearch. [License from](https://github.com/devMagno/ui-clone-google)

#### Presearch-YoutubeSearch-theme *after*
![image](https://user-images.githubusercontent.com/111701513/207109308-ddd1a8ce-2d1b-4141-bfe2-79eb680bfe60.png)

**image-description:** As we can see in the image, "I thought about improving the presearch search interface". This is a YoutubeSearch custom theme demo for Presearch. [License from](https://dribbble.com/shots/15036214-Youtube-Redesign-Concept-Dark-theme-Part-2)

#### why?
- Before coding, it is necessary to evaluate whether the idea is good or bad
- I was thinking about creating custom themes for the pre-search
- Presearch can be better known using custom themes or based on themes... in existing solutions

#### question/feedback
*What do you all think of this idea?*

**Comment by ghost** [comment]
Hi everyone.

#### concept
*A random idea I had today. Would it be possible to index any site with static content through a cdn in presearch?*

#### idea
```html
<img src="http://cdn.presearch.io/?u=http://domain-website.com/logo.gif"/>
```

**Comment by jejopl** [comment]
Hey guys, sorry I didn't have much time recently for you. Will check those out shortly, thanks for the ideas!

**Reply by ghost** [reply]
Hi jejopl.

> Hey guys, sorry I didn't have much time recently for you. Will check those out shortly, thanks for the ideas!

thank you for feedback.

**Comment by silicology1** [comment]
[Souct](https://scout.click) is really great to improve quality results. But also include [brave goggle feature](https://search.brave.com/help/goggles).

**Comment by silicology1** [comment]
Just don't like the ads model.  Some thing like giving back the pre tokens earned to get ads free experience. Some monthly quota with which one can browse ad free.
